Src/Newtonsoft.Json/Utilities/JavaScriptUtils.cs

```diff
@@ -314,5 +313,175 @@ public static string ToEscapedJavaScriptString(string value, char delimiter, boo
+                        default:
+                            if (c < charEscapeFlags.Length || stringEscapeHandling == StringEscapeHandling.EscapeNonAscii)
+                            {
+                                if (c == '\'' && stringEscapeHandling != StringEscapeHandling.EscapeHtml)
+                                {
+                                    escapedValue = @"\'";
+                                }
+                                else if (c == '"' && stringEscapeHandling != StringEscapeHandling.EscapeHtml)
+                                {
+                                    escapedValue = @"\""";
+                                }
+                                else
+                                {
+                                    if (writeBuffer == null || writeBuffer.Length < UnicodeTextLength)
+                                    {
+                                        writeBuffer = BufferUtils.EnsureBufferSize(bufferPool, UnicodeTextLength, writeBuffer);
+                                    }
+
+                                    StringUtils.ToCharAsUnicode(c, writeBuffer);
+
+                                    // slightly hacky but it saves multiple conditions in if test
+                                    escapedValue = EscapedUnicodeText;
+                                }
+                            }
+                            else
+                            {
+                                escapedValue = null;
+                            }
+                            break;
+                    }
+
+                    if (escapedValue == null)
+                    {
+                        continue;
+                    }
+
+                    bool isEscapedUnicodeText = string.Equals(escapedValue, EscapedUnicodeText);
+
+                    if (i > lastWritePosition)
+                    {
+                        int length = i - lastWritePosition + ((isEscapedUnicodeText) ? UnicodeTextLength : 0);
+                        int start = (isEscapedUnicodeText) ? UnicodeTextLength : 0;
+
+                        if (writeBuffer == null || writeBuffer.Length < length)
+                        {
+                            char[] newBuffer = BufferUtils.RentBuffer(bufferPool, length);
+
+                            // the unicode text is already in the buffer
+                            // copy it over when creating new buffer
+                            if (isEscapedUnicodeText)
+                            {
+                                Array.Copy(writeBuffer, newBuffer, UnicodeTextLength);
+                            }
+
+                            BufferUtils.ReturnBuffer(bufferPool, writeBuffer);
+
+                            writeBuffer = newBuffer;
+                        }
+
+                        s.CopyTo(lastWritePosition, writeBuffer, start, length - start);
+
+                        cancellationToken.ThrowIfCancellationRequested();
```

Comment:
It's subject to debate, but I think reasonable guidelines are:

1. Avoid doing anything with side-effects before at least one check.
2. On the other hand, avoid doing a check just before a call that will itself check.
3. Avoid large allocations or anything "weighty" before at least one check. (Though in one of the revision commits I removed a few I was doing before `EnsureWriteBuffer` on the basis that since that allocation is cached and comes from a shared store of buffers it's not quite as "heavy" as if it was a one-off allocation of a similarly-sized array)
4. Avoid having loops with no checks within them (but again, if you're calling a method that will itself check in that loop, that's fine.
5. It's better to check before an operation than after it.

Throwing within an `async` method results in the `Task` returned by that method having `IsCanceled` as true. Conversely, `await`ing a cancelled task results in a throw of a `TaskCanceledException` and calling `Wait()` or `Result` on it in an `AggregateException` wrapping one.

As such, throw when the method is `async` return a cancelled task when it isn't.